tests/host_tools/uffd/src/bin/valid_handler.rs

```diff
@@ -0,0 +1,289 @@
+    }
+
+    fn update_mem_state_mappings(&mut self, start: u64, end: u64, state: &MemPageState) {
+        for region in self.mem_regions.iter_mut() {
+            for (key, value) in region.page_states.iter_mut() {
+                if key >= &start && key < &end {
+                    *value = state.clone();
+                }
+            }
+        }
+    }
+
+    fn populate_from_file(&self, region: &MemRegion) -> (u64, u64) {
+        let src = self.backing_buffer as u64 + region.mapping.offset;
+        let start_addr = region.mapping.base_host_virt_addr;
+        let len = region.mapping.size;
+        // Populate whole region from backing mem-file.
+        // This offers an example of how memory can be loaded in RAM,
+        // however this can be adjusted to accommodate use case needs.
+        let ret = unsafe {
+            self.uffd
+                .copy(src as *const _, start_addr as *mut _, len, true)
+                .expect("Uffd copy failed")
+        };
+
+        // Make sure the UFFD copied some bytes.
+        assert!(ret > 0);
+
+        return (start_addr, start_addr + len as u64);
+    }
+
+    fn zero_out(&mut self, addr: u64) -> (u64, u64) {
+        let page_size = get_page_size().unwrap();
+
+        let ret = unsafe {
+            self.uffd
+                .zeropage(addr as *mut _, page_size, true)
+                .expect("Uffd zeropage failed")
+        };
+        // Make sure the UFFD zeroed out some bytes.
+        assert!(ret > 0);
+
+        return (addr, addr + page_size as u64);
+    }
+
+    fn serve_pf(&mut self, addr: *mut u8) {
+        let page_size = get_page_size().unwrap();
+
+        // Find the start of the page that the current faulting address belongs to.
+        let dst = (addr as usize & !(page_size as usize - 1)) as *mut c_void;
+        let fault_page_addr = dst as u64;
+
+        // Get the state of the current faulting page.
+        for region in self.mem_regions.iter() {
+            match region.page_states.get(&fault_page_addr) {
+                // Our simple PF handler has a simple strategy:
+                // There exist 4 states in which a memory page can be in:
+                // 1. Uninitialized - page was never touched
+                // 2. FromFile - the page is populated with content from snapshotted memory file
+                // 3. Removed - MADV_DONTNEED was called due to balloon inflation
+                // 4. Anonymous - page was zeroed out -> this implies that more than one page fault
+                //    event was received. This can be a consequence of guest reclaiming back its
+                //    memory from the host (through balloon device)
+                Some(MemPageState::Uninitialized) => {
+                    let (start, end) = self.populate_from_file(region);
+                    self.update_mem_state_mappings(start, end, &MemPageState::FromFile);
+                    return;
+                }
+                Some(MemPageState::FromFile)
```

Comment:
so if we get a page fault for a page that we've already faulted in from the file, we will zero it?
shouldn't it be a no-op?

Reply:
I don't think it matters much how we handle `MemPageState::FromFile` pages in our case, since our test does not trigger this path (page faulting a memory region that has been previously loaded into memory).

Reply:
what if we page in a region from a file, and the guest OS swaps it to disk. The guest then wants to access it again and it's not in memory. How will that be handled?

Reply:
No-op would not help in this scenario. If the guest swaps a page that is in the `FromFile` state, and then a page fault occurs on that page, the UFFD PF handler should load the page into memory again.

Our test case does not trigger this scenario, but I will update the handler accordingly for completeness.